### ISSUE TITLE
Speed Ez - Taxes by Child Account

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1293]
In GitLab by @kdjstudios on Dec 19, 2018, 15:38

**Submitted by:**  "Kyle Johnson" <kyle.johnson@sahosted.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2018-12-18-19662/conversation

**Gary HD:** http://www.servicedesk.answernet.com/profiles/ticket/6339665

**Server:** ALL

**Client/Site:** ALL

**Account:** ALL

**Issue:** 

I was just on the phone with Jo Ann and she mentioned that she would like to see the taxes broke down on a per child account basis. Currently it appears that all taxes are being summed into one line at the bottom. However, this is leading to each child account not paying their portion of the taxes. May we please confirm if this is currently possible. If so, how does one set this up. If not, may we have an estimate on what it would take to be bale to accomplish such functionality.

Original ticket #270

`Test Cases:` https://docs.google.com/spreadsheets/d/1w8JPABnQTYm0GCgF4V71X2ZBLF-ZbdUic_55GtIT0ak/edit?usp=sharing

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 19, 2018, 16:05

@smasih @amishra @pchaudhary @tim.traylor 

Can we please get an estimate on this? Gary also wrote inquiring that he thought this was already possible.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 20, 2018, 08:29

@smasih @amishra @pchaudhary @tim.traylor @abakshi 

Gary wrote:

> That makes no sense to me.  We are simply doing separate bills and then combining them so why wouldn't all calculations be available on the individual accounts?  This is true where individual accounts my have different taxes.
> 
> Please give me a time estimate on this
> 
> We are going to have to make it optional if we do it as we don't want to change every account

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 20, 2018, 08:41

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 20, 2018, 10:02

@amishra @smasih @pchaudhary @tim.traylor 

We will need to have an estimate on this to give to Gary. This function will need to be something that can be enabled or disabled.

Tim wrote to Gary with an 'random' guesstimate of 4-6 hours.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 21, 2018, 03:17

@kdjstudios 

This is a core issue. I checked this is not a printing fault.
  
This is going to have an impact on the invoice functionality so we need to do a lot of testing.

* Negative tax
* positive tax
* checking total
* Reversals(drafts)
* Billing Cycle reversals


|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-2|6|Aware of Concepts, implementation tricky|10.5|High Impact lots of use cases|10.5|21|

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 21, 2018, 08:14

I wrote:

> The support team was able to look further into this and found that that this not a printing fault; which I believe is what Tim was thinking when he provided his estimate. It actually impacts a larger area.This is going to have an impact on core items of invoice functionality, listed below.
> 
> Areas of impact:
> - Negative tax
> - Positive tax
> - Checking total
> - Reversals(drafts)
> - Billing Cycle reversals
> 
> Estimates:
> 
> Lower Time Bound: 6 hours
> Upper Time Bound: 10.5 hours
> Time for Testing: 10.5
> 
> Overall Max Time for Dev & Testing: 21 Hours
> 
> Now that we have an estimate provided, we will still need to know if this is something we are considering a Bug and are fixing internally. Or if this is an enchantment and we will need to notify Dave to reach out to Jo Ann about billing charges for this?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 27, 2018, 15:06

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 14, 2019, 09:28

I wrote:

> The support team was able to research and confirm that currently SA Billing is setup to only charge the taxes at the parent account level. They believe it is possibly to develop the functionality to allow for the taxes by child account and are currently looking into what all it will take to allow for having separate line items per child account per tax. We will keep you informed with any progress on this topic.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 14, 2019, 11:14

Jo Ann wrote:

> It also does not charge Holiday charges per child account.

I opened a new ticket to address this separate issue: #1327

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 15, 2019, 02:50

@kdjstudios 

* New estimate after getting new scope added on the taxes part, now the **holiday charges** (Issue #1327 ) are also should be handled at the same time. 

|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Large-1|8|Aware of Concepts, implementation tricky|14|High Impact lots of use cases|14.0|28|

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 16, 2019, 10:16

Just adding another test case from #1326 :

> Each account should only charge applicable taxes to its account. Such if a child account is setup for taxes and the parent account is not. there should still be the child account taxes on the parent account's invoice, but only on the child account fees.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 17, 2019, 10:05

This was partially discussed today, Gary said this is how the system should work. Marking for Discussion.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 29, 2019, 12:21

@smasih @amishra @pchaudhary @tim.traylor 

Sorry I miss labeled this. Since this was already discussed and Gary stated this is how the system should work this should have been marked for development. Updating label.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jan 30, 2019, 21:37

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 31, 2019, 04:54

@kdjstudios 

Thank you for the update.

Here we have described testing scenarios of the taxes.

| Scenarios                        | Parent(p1) | Child(c1) | Expected Result                                                                     | System Result              |
|----------------------------------|------------|-----------|-------------------------------------------------------------------------------------|----------------------------|
| If p1 has a tax and c1 has a tax | 10          | 2         | | sum of the all line item(p1+c1)*10/100 |
| If only p1 has a tax             | 1          | 0         |          | sum of the all line item(p1+c1)*1/100 |
| If the only c1 has a tax             | 0          | 3         |       | sum of the all line item(p1+c1) |
| If no taxes                      | 0          | 0         | ((Sum of the p1 line item amount) + (Sum of the c1 line item amount))               | sum of the all line item(p1+c1) |




Before starting the development we have a question on the above first three scenarios.

**Example:**

If p1 has amount $100 and tax 2% and c1 has amount $150 and tax 5%.   

1.As per the current code the system calculate with **(p1+p2)*2/100.**  
2.As we thing this should be calculated **(p1*2/100)+(c1*5/100).**

**Which should one correct?**

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 31, 2019, 12:17

@amishra @smasih @pchaudhary 

Taxes need to be calculated at the account level according to the accounts settings independently. Then if the account is specified to 'Bill To' then they would sum the same taxes together, but different taxes would still be separate. If the account is specified to 'separate child account transactions' then each account will need to have its own tax in its account's group of charges. Please let me know if any further explanation is needed.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2019, 11:27

@amishra @smasih @pchaudhary 

Has any progress been made on this? I also beleive we are addressing #1327 on this ticket too correct?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 7, 2019, 02:41

@kdjstudios 

Thank you for the update.

Please explain what do you understand by **Bill To** and **separate child account transactions**.

We explain the **Bill To** and **Separate child account transactions** functionality of the current code.

**Bill To**

* We have one parent account **p1** and child account **c1**.
* p1 has 4 account activity and c1 has 5 account activity.
* If **Bill To** is **true** then iterate over **p1** account activities.
* If **Bill To** is **false** then iterate over  **p1** and its **c1**(active child account) account activities.

**Separate child account transactions**  

* Separate child account transactions are only used when downloading receipt from the client portal.


Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Feb 11, 2019, 15:44

@kdjstudios 

seems like current "bill to" functionality might have some bug in it so we need to know what is yours/ops understanding with bill to so that we can confirm its working as expected. Because "bill_to" understanding is crucial for moving further on this ticket.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 12, 2019, 10:00

@smasih @amishra @pchaudhary @tim.traylor 

Please look at #1362 for explanations of the options and example configurations.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Feb 13, 2019, 24:24

Tested 

Two Accounts
Parent P1 with one Usage Activity and one Recurring $10   
Child C1 with one Usage Activity and one Recurring $20

With Bill to False
![withBillToFalse](/uploads/8a6850c9a5ae501d817944592ef037ad/withBillToFalse.png)

With Bill to True
![withBillToTrue](/uploads/cb1ce3601c3c87d1a590d7b280b38ae9/withBillToTrue.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 13, 2019, 10:57

@smasih 

I just want to confirm that Separate Child Transactions was also enabled in both of the above examples. If so, then this appears to be billing correctly.

EDIT: I am a bit concerned with the above examples. This ticket is about resolving the issue with Taxes on child accounts, and there are no tax charges on either of the examples above, so I am unsure if they are a good representation of what we should be looking at.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 18, 2019, 04:50

@kdjstudios 

Thank you for the update



> I just want to confirm that Separate Child Transactions was also enabled in both of the above examples. If so, then this appears to be billing correctly.

Here are the examples with **Separate Child Transactions** enabled.
* **Bill To true**

![merge_bill_true](/uploads/1537db88ea9705f8a4b8500ea8748b66/merge_bill_true.png)

* **Bill to False**
![merge_bill_to_false](/uploads/36bb406eedaaf2d3750640bcf062d391/merge_bill_to_false.png)




> EDIT: I am a bit concerned with the above examples. This ticket is about resolving the issue with Taxes on child accounts, and there are no tax charges on either of the examples above, so I am unsure if they are a good representation of what we should be looking at.

Currently, we are trying to understand the existing **bill to** and **Separate Child Transactions** functionality before we started.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 18, 2019, 12:03

@amishra 

Please read #1362 for a complete explanation on how those features function and post any questions and updates there. Please keep this ticket in regards to having taxes on child account separate out, just like billing charges do when the 'separate child transactions' is enabled.

In additional clarification too. The above examples appear are showing correctly how the bill to and separate child transactions should work, and I presume have always worked; except for the major issue of Taxes not being separated out still. Also does not appear that http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1296#note_34401 in regards to holiday fees may still not have been developed.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Feb 19, 2019, 06:53

@kdjstudios 


Thank you for the update.


As per my understanding the current functionality of the **separate child transactions** doesn't work on the current code of SAB, which means there is no impact either we make it true or false, the line items don't show any impact accordingly.  


According to your comment #1362, we need to change the code:  

* If we make a single line item of the child accounts, (*when the separate child transactions is false*) then the option to change the base rate & quantity for the child account's line item is not possible.  

Please allow if we still need to make those changes.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 19, 2019, 09:05

@amishra @smasih @pchaudhary @tim.traylor 

1. Where do I mention we need to change code? What code needs to be changed? Why does the code need to be changed? Please provide examples and explanations on the setup you performed and how the results did not match the setup. NOTE: your examples above lack detail to understand what you performed; and what your expected results are; and what the actual results were. Please feel free to use the same test case scenario document template we are using in #28 for testing. 
2. What is your understanding of the separate child transactions? How does this differ from what is in #1362 ?

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Feb 20, 2019, 24:33

@kdjstudios 

Thank you for the update & please find the update to all your concern below.  


> Where do I mention we need to change the code? 

* As we already mentioned that on the ticket #1362, you pointed out the flow that how should Separate child account transactions should work :
  * Separate child account transactions – This option allows for the invoice to show each child accounts usage on separate line items.   
      - When this is checked each child accounts line items will be shown.  
      - When this is not checked they will be combined for a type of line item. 

* According to us which means that if the Separate child account transaction is checked then the line items of the child account will be shown as it is but if it is unchecked then the child account's line items (if more than 1 line item) would be merged as a single line item. 
 
> Why does the code need to be changed?

* We had already mentioned on our last comment that as per the current SAB code the Separate child account transactions do not have any impact on the invoice, which means there'll be no impact either we make it true or false.  


> Please provide examples and explanations on the setup you performed and how the results did not match the setup.

* Please find the below example in which **DM-AC-1** is the parent account & **DEMO-54** is the child.  
  * Either the Separate child account transaction is True or False, the draft invoice shown the same number of line items.  
  * ![image](/uploads/8199fc7b38effa7ee43b5fa0b38e5829/image.png)

  * And according to your comment on the #1362, the line items of **DEMO-54** will be merged as a single line item showing the **Account Name on the description & $700 as the base rate & amount**.  

> What is your understanding of the separate child transactions? How does this differ from what is in #1362 ?

* We just want to confirm that the separate child acocunt transaction check box do not have any impact on the current format of invoice, either it's true or False.

Hope this helps.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 20, 2019, 09:51

@smasih @pchaudhary @amishra @tim.traylor 

Moving the discussion once again to the correct ticket. http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1296#note_35306 Please post all questions and concerns in #1362 about how the parent child and their configuration options work. 

Please keep only posts regarding taxes separated out on child account in this ticket. I have yet to see one screen shot or any testing showing any development on separating taxes out on child accounts? How much time has been spent on this, but no progress has been made so far?

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Mar 4, 2019, 24:03

Hi [Kyle] ( @kdjstudios ),  

#1296 , #1327 & #1362 are linked with each other & dependent too. But the label are totally differents from each other. 

We request you to sync the labels & priorities so that it will be easy to handle.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 4, 2019, 09:47

Placing in pause until #1362 is completed. Need that working in order to have this working.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 27, 2019, 12:33

Moving this back to development.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 18, 2019, 01:40

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 18, 2019, 04:09

@kdjstudios @jredfield 

Thank you for the update.

Before starting the development we want to present the current functionality of the taxes.
So, We have described current functionality in the [Sheet  mentioned on the top.](https://docs.google.com/spreadsheets/d/1w8JPABnQTYm0GCgF4V71X2ZBLF-ZbdUic_55GtIT0ak/edit?usp=sharing)
Please have a look and fill the expected functionality of the column. It'll help us to develop the functionality according to your expectations.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 18, 2019, 04:11

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 21, 2019, 09:18

@kdjstudios 





The sheet linked below has the current functionality of calculating taxes, we will need Answernet team to fill up the expected functionality.   

You might need to go through [this sheet](https://docs.google.com/spreadsheets/d/1w8JPABnQTYm0GCgF4V71X2ZBLF-ZbdUic_55GtIT0ak/edit?usp=sharing) before pulling this ticket up for Tuesday Meeting.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 21, 2019, 14:15

@amishra @smasih @jredfield 

I am confused with why we are needing to fill out a spreadsheet? This needs to behave in the same exact manor that charges on the accounts are handled, now that #1362 has already detailed this out and fixed it?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 24, 2019, 05:41

@kdjstudios @amishra 

>> From 1362: Usage-based charges, Late Fees, Holidays, and Taxes will need to be billable items that are affected by the "bill to" and "separate transactions"; while Base Rates, Additional Fees, and Staged Fees will not be affected.

That is all that is explained for taxes in the #1362 and so we not only need to know how taxes should be calculated we also need to know how they will be displayed.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 24, 2019, 09:14

This was discussed today and as pointed out we will want the taxes to be handled exactly like the charges are being handled on the #1362

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 1, 2019, 07:54

@kdjstudios @jredfield 

**Question:**

If "Separate Account Transaction is turned OFF" then `Sum of Taxable Lineitems (parent + child)` with `Parent Tax` is used to calculate total tax.

**Do we need to keep it that way?** or do we need to use the `parent tax` against parent lineitems and `child tax` against child lineitems.

Please let me know if not clear

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 4, 2019, 13:48

@amishra 

Where did you find the following. I am not seeing that in the help documentation. "If "Separate Account Transaction is turned OFF" then `Sum of Taxable Lineitems (parent + child)` with `Parent Tax` is used to calculate total tax."

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 5, 2019, 01:11

@kdjstudios 

> Where did you find the following. I am not seeing that in the help documentation. "If "Separate Account Transaction is turned OFF" then `Sum of Taxable Lineitems (parent + child)` with `Parent Tax` is used to calculate total tax."


* Not sure, if it's there on the Help Documentation I just explained the current functionality of SAB in relation to TAX calculation and have asked a question if we need to change it.

* This question is acting as a roadblock and we are not able to develop further on this ticket.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 5, 2019, 10:50

@amishra @smasih @jredfield 

Please use the updates in #1362 to know when and how to assess the fees based upon the options selected. If you have questions please refer to the help documentation. Please note your team should have this all properly documented in the tech documents also.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Nov 6, 2019, 03:34

@kdjstudios @jredfield @amishra 

That's the thing these were never defined therefore its not anywhere in the documentation.

Requirement on the ticket is clear and we are supposed to do it as discussed in #1362 **but what we don't know is the bifurcation of taxes.** 

We can only show taxes in account groups if we know the bifurcation of the total tax, and that is exactly what we have been asking for. 

Let's pick this up on Thursday call  


Updating Discussion label

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 7, 2019, 09:10

This was discussed today, we want the taxes to be charged in the same way we apply usage charges in #1362

We will need to have an update sent to operations. For them to check all of the taxes on their parent/child accounts.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 7, 2019, 11:04

@rchauhan @smasih @amishra @jredfield 

It was mentioned in WhatsApp about taxes applied based upon state. This is in full control by operations. They can add as many tax rates at the site level and then can apply each appropriately to the account accordingly. They also have the ability to apply more then one tax at the account level if needed to. Lets say a County, state, and federal tax, etc.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 13, 2019, 04:05

@kdjstudios @jredfield 

Thank you for the update.

We have completed the development and setup branch on the merge2 server for your testing. Please test it and give me your valuable feedback.  

Our Test cases mentioned [Here](http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/Taxes-Test-Cases-For-Ticket-%231296)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 14, 2019, 13:41

@kdjstudios @jredfield 

I request you please test this so that we can set up another branch for your testing.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2019, 14:50

@amishra @jredfield 

The test cases and results in http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/Taxes-Test-Cases-For-Ticket-%231296 appear to be correct. Marking this for announcement.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 23, 2019, 05:01

@kdjstudios @jredfield 

Thank you for the update.

We have successfully released on both of the production servers.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 23, 2019, 11:53

The taxes are being applied correctly

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 4, 2019, 12:35

closed
